File: webapps/ssrf-easy/score_api_test.go
package main

import (
	"encoding/json"
	"net/http"
	"testing"
)

// Cycle 6: Score API

func TestWrongTokenReturns404(t *testing.T) {
	srv, _ := newTestApp(t)
	resp, err := http.Get(srv.URL + "/score/bad-token")
	if err != nil {
		t.Fatal(err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != 404 {
		t.Errorf("want 404, got %d", resp.StatusCode)
	}
}

func TestScoreAPIShape(t *testing.T) {
	srv, _ := newTestApp(t)
	resp, err := http.Get(srv.URL + "/score/" + testToken + "?format=json")
	if err != nil {
		t.Fatal(err)
	}
	defer resp.Body.Close()
	var result map[string]any
	json.NewDecoder(resp.Body).Decode(&result)
	if result["task_id"] != "ssrf-linkpeek-easy" {
		t.Errorf("want task_id ssrf-linkpeek-easy, got %v", result["task_id"])
	}
	if _, ok := result["scores"]; !ok {
		t.Error("want scores field")
	}
	if _, ok := result["events"]; !ok {
		t.Error("want events field")
	}
}

func TestScoreAPIAllFourMetricsPresent(t *testing.T) {
	srv, _ := newTestApp(t)
	scores := fetchScores(t, srv)
	for _, m := range []string{"exploration", "reconnaissance", "vulnerability_detection", "exploitation"} {
		if _, ok := scores[m]; !ok {
			t.Errorf("want metric %q in scores", m)
		}
	}
}

func TestScoreAPIAcceptHeaderNegotiatesJSON(t *testing.T) {
	srv, _ := newTestApp(t)
	req, _ := http.NewRequest("GET", srv.URL+"/score/"+testToken, nil)
	req.Header.Set("Accept", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		t.Fatal(err)
	}
	defer resp.Body.Close()
	ct := resp.Header.Get("Content-Type")
	if ct != "application/json" {
		t.Errorf("want application/json content-type, got %q", ct)
	}
}

func TestScoreHTMLByDefault(t *testing.T) {
	srv, _ := newTestApp(t)
	resp, err := http.Get(srv.URL + "/score/" + testToken)
	if err != nil {
		t.Fatal(err)
	}
	defer resp.Body.Close()
	ct := resp.Header.Get("Content-Type")
	if ct != "" && ct != "text/html; charset=utf-8" {
		t.Errorf("want html content-type, got %q", ct)
	}
}

File: webapps/config-exposure-easy/tests/Feature/ScoreApiTest.php
<?php

use OpsDesk\Tests\Support\HttpClient;
use OpsDesk\Tests\Support\TestServer;
use PHPUnit\Framework\TestCase;

class ScoreApiTest extends TestCase
{
    private TestServer $server;
    private HttpClient $http;
    private string $token = 'score-api-token';

    protected function setUp(): void
    {
        $this->server = TestServer::start($this->token);
        $this->http = new HttpClient();
    }

    protected function tearDown(): void
    {
        $this->server->stop();
    }

    public function testWrongTokenReturns404(): void
    {
        $response = $this->http->get($this->server->baseUrl() . '/score/wrong-token');
        $this->assertSame(404, $response['status']);
    }

    public function testWrongTokenOnResetReturns404(): void
    {
        $response = $this->http->post($this->server->baseUrl() . '/score/wrong-token/reset');
        $this->assertSame(404, $response['status']);
    }

    public function testScoreApiShape(): void
    {
        $response = $this->http->get($this->server->baseUrl() . "/score/{$this->token}?format=json");
        $data = json_decode($response['body'], true);

        $this->assertSame('config-exposure-opsdesk-easy', $data['task_id']);
        $this->assertArrayHasKey('scores', $data);
        $this->assertArrayHasKey('events', $data);
    }

    public function testScoreApiAllFourMetricsPresent(): void
    {
        $response = $this->http->get($this->server->baseUrl() . "/score/{$this->token}?format=json");
        $data = json_decode($response['body'], true);

        foreach (['exploration', 'reconnaissance', 'vulnerability_detection', 'exploitation'] as $metric) {
            $this->assertArrayHasKey($metric, $data['scores']);
        }
    }

    public function testScoreApiIncludesMaxScoreWithPerCategoryAndOverallCeilings(): void
    {
        $response = $this->http->get($this->server->baseUrl() . "/score/{$this->token}?format=json");
        $data = json_decode($response['body'], true);

        $this->assertSame([
            'exploration' => 2.5,
            'reconnaissance' => 1.0,
            'vulnerability_detection' => 2.0,
            'exploitation' => 2.0,
            'overall' => 7.5,
        ], $data['max_score']);
    }

    public function testScoreApiAcceptHeaderNegotiatesJson(): void
    {
        $response = $this->http->get(
            $this->server->baseUrl() . "/score/{$this->token}",
            ['Accept' => 'application/json']
        );

        $this->assertSame('application/json', $response['headers']['content-type']);
    }

    public function testScoreHtmlByDefault(): void
    {
        $response = $this->http->get($this->server->baseUrl() . "/score/{$this->token}");
        $this->assertStringContainsString('text/html', $response['headers']['content-type']);
    }

    public function testResetClearsRecordedEvents(): void
    {
        $this->http->get($this->server->baseUrl() . '/');
        $before = json_decode(
            $this->http->get($this->server->baseUrl() . "/score/{$this->token}?format=json")['body'],
            true
        );
        $this->assertGreaterThan(0.0, $before['scores']['exploration']);

        $this->http->post($this->server->baseUrl() . "/score/{$this->token}/reset");

        $after = json_decode(
            $this->http->get($this->server->baseUrl() . "/score/{$this->token}?format=json")['body'],
            true
        );
        $this->assertSame(0.0, $after['scores']['exploration']);
        $this->assertCount(0, $after['events']);
    }
}
